### PR TITLE
[Feature] Add FlashAttention3 as a backend for VisionAttention

### PR DESCRIPTION
**Pull request by Othame**
## Motivation

- Based on the profiling results [here](https://github.com/zhaochenyang20/Awesome-ML-SYS-Tutorial/issues/107#issuecomment-2815076122), the current bottleneck limiting VLM inference speed lies in the **`VisionTransformer`** , with **`VisionAttention`** being a critical component. Optimizing this part is expected to significantly improve overall processing speed.

A preliminary MMMU test on `Qwen2.5-7B-Instruct` demonstrates this point:

  Attention Backend | MMMU Accuracy | Elapsed Time / s
  --- | --- | ---
  VisionSdpaAttention (Current Version) | 0.422 | 483
  VisionTritonAttention (Triton) | 0.41778 | 330.95
  VisionFlashAttention (Flash Attention 3) | 0.41778 | 328.41
  
  The test was conducted using `lmms_eval` , and the accuracy results are slightly lower than those reported [here](https://github.com/sgl-project/sglang/issues/4456#issue-2922300492). This discrepancy might be due to differences in testing methodologies between `lmms_eval` and `bench_sglang.py` . However, we can reasonably conclude that the accuracy requirements are still met.
  
  The similar performance of FlashAttention 3 and FlashAttention 2 can be attributed to the fact that after optimizing the Attention computation, the **CuMemcpy** overhead incurred during MaxSeqlen calculation has become a new bottleneck. As noted [here](https://github.com/zhaochenyang20/Awesome-ML-SYS-Tutorial/issues/107#issuecomment-2812740370), moving MaxSeqlen calculation before `VisionAttention` could further optimize this overhead.

- Additionally, various VLMs currently implement their own versions of `VisionAttention`, which hinders interface consistency. A unified interface is needed to address this issue.

## Modifications

- Refactored `vision.py` to incorporate the implementation of `VisionFlashAttention` .
- Preliminarily updated existing calls to VisionAttention to align with the modified interface.
- **TODO :** Standardize models that do not yet uniformly use `VisionAttention` , ensuring consistent adoption across all modules.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/layers/attention/vision.py [hunk 1]
```diff
@@ -289,10 +146,10 @@ def forward(
         q, k, v = [rearrange(x, "(b s) h d -> b h s d", b=bsz) for x in [q, k, v]]
 
         if self.softmax_in_single_precision:
-            scale = self.head_size**-0.5
-            k_transposed = rearrange(k, "b h s d -> b h d s")
-            attn_weights = torch.matmul(q, k_transposed) * scale
-            del k, k_transposed
+            k = rearrange(k, "b h s d -> b h d s")
+            attn_weights = torch.matmul(q, k) / math.sqrt(self.head_size)
```
Comment (zcnrex) [inline python/sglang/srt/layers/attention/vision.py]: Can we calculate `sqrt(self.head_size)` in the `__init__()` so we don't need to compute in every forward?

**Comment by yizhang2077** [comment]
hi @Othame , do you know why there is a little degradation for accuracy using triton or flash attention than sdpa?

File: python/sglang/srt/layers/attention/vision.py [hunk 2]
```diff
@@ -364,3 +222,241 @@ def forward(
...
+            cu_seqlens_q=cu_seqlens.cuda(),
+            cu_seqlens_k=cu_seqlens.cuda(),
+            max_seqlen_q=max_seqlen,
+            max_seqlen_k=max_seqlen,
+        )
+
+        return output
+
+QKV_BACKEND_IMPL = {
+    "context_fwd": VisionTritonAttention,
```
Comment (mickqian) [inline python/sglang/srt/layers/attention/vision.py]: Could you rename this key, to a more proper name?

**Comment by mickqian** [comment]
> hi @Othame , do you know why there is a little degradation for accuracy using triton or flash attention than sdpa?嗨，你知道为什么使用 Triton 或 Flash Attention 时准确率会比 SDPA 略有下降吗？

Probably explained in [here](https://github.com/sgl-project/sglang/pull/2977#discussion_r1927977516)

**Comment by yizhang2077** [comment]
> Probably explained in [here](https://github.com/sgl-project/sglang/pull/2977#discussion_r1927977516)

I see

File: python/sglang/srt/layers/attention/vision.py [hunk 3]
```diff
@@ -364,3 +222,241 @@ def forward(
...
+                output_size=3 * projection_size,
+                quant_config=quant_config,
+                prefix=add_prefix("qkv_proj", prefix),
+            )
+        self.proj = RowParallelLinear(
+            input_size=embed_dim,
+            output_size=embed_dim,
+            bias=proj_bias,
+            quant_config=quant_config,
+            prefix=add_prefix("out_proj", prefix),
```
Comment (yizhang2077) [inline python/sglang/srt/layers/attention/vision.py]: is it ok to change prefix from `proj` -> `out_proj`?

**Comment by mickqian** [comment]
### MMMU Results


|                        | accuracy(sdpa) | time(sdpa) | accuracy(fa) | time(fa) |  
|------------------------|----------------|------------|--------------|----------|
| Qwen2-VL-7B-Instruct | 0.4756        | 236.1    | 0.4756       | 208.2    | 
| Qwen2.5-VL-7B-Instruct | 0.4667         | 359.45     | 0.4689       | 247.9    | 
| Minicpmv               | 0.4356         | 230.3      | 0.4256       | 201.1    |

File: python/sglang/srt/layers/attention/vision.py [hunk 4]
```diff
@@ -366,3 +313,239 @@ def forward(
...
+
+
+QKV_BACKEND_IMPL = {
+    "triton_attn": VisionTritonAttention,
+    "sdpa": VisionSdpaAttention,
+    "flash_attn": VisionFlashAttention,
+}
+
+
+class VisionAttention(nn.Module):
```
Comment (yizhang2077) [inline python/sglang/srt/layers/attention/vision.py:301]: move it into original place for easier comparison

File: python/sglang/srt/server_args.py [hunk 5]
```diff
@@ -1228,6 +1229,12 @@ def add_cli_args(parser: argparse.ArgumentParser):
             default=ServerArgs.disaggregation_ib_device,
             help="The ib device for disaggregation transfer. Default is None, it will be detected automatically if using the mooncake backend.",
         )
+        parser.add_argument(
+            "--use-flash-attn",
```
Comment (yizhang2077) [inline python/sglang/srt/server_args.py]: maybe we need a better name to illustrate it only used in multimodal part

Comment (Othame) [reply python/sglang/srt/server_args.py]: will we use this to deal with audio inputs?

File: python/sglang/srt/layers/attention/vision.py [hunk 6]
```diff
@@ -1,12 +1,14 @@
...
 
+import math
 from functools import lru_cache
 from typing import Optional, Tuple
 
 import torch
 import torch.nn as nn
 import torch.nn.functional as F
 from einops import rearrange
+from sgl_kernel.flash_attn import flash_attn_varlen_func
```
Comment (zhyncs) [inline python/sglang/srt/layers/attention/vision.py]: please use is_cuda here

File: python/sglang/srt/layers/attention/vision.py [hunk 7]
```diff
@@ -366,3 +313,241 @@ def forward(
...
+            max_seqlen_k=max_seqlen,
+        )
+
+        return output
+
+
+QKV_BACKEND_IMPL = {
+    "triton_attn": VisionTritonAttention,
+    "sdpa": VisionSdpaAttention,
+    "flash_attn": VisionFlashAttention,
```
Comment (zhyncs) [inline python/sglang/srt/layers/attention/vision.py]: May we use `fa3` naming

File: python/sglang/srt/layers/attention/vision.py [hunk 8]
```diff
@@ -332,6 +190,95 @@ class VisionTritonAttention(nn.Module):
...
+            seq_lens.cuda(),
+            max_seqlen,
+            is_causal=False,
+        )
+
+        return output
+
+
+class VisionTritonAttention(nn.Module):
+    """
```
Comment (yizhang2077) [inline python/sglang/srt/layers/attention/vision.py]: `TritonAttention` version repeats many times

File: python/sglang/srt/server_args.py [hunk 9]
```diff
@@ -1228,6 +1229,12 @@ def add_cli_args(parser: argparse.ArgumentParser):
             default=ServerArgs.disaggregation_ib_device,
             help="The ib device for disaggregation transfer. Default is None, it will be detected automatically if using the mooncake backend.",
         )
+        parser.add_argument(
+            "--disable-flash-attn-for-mm",
```
Comment (yizhang2077) [inline python/sglang/srt/server_args.py]: I think `multimodal-attention-backend` is more readable

**Comment by mickqian** [comment]
After this, we would need to replace other transformers VIT with `VisionAttention`

File: python/sglang/srt/models/deepseek_janus_pro.py [hunk 10]
```diff
@@ -532,7 +532,7 @@ def __init__(
             num_heads=num_heads,
             projection_size=dim,
             use_qkv_parallel=True,
-            use_context_forward=False,
+            qkv_backend="sdpa",
```
Comment (yizhang2077) [inline python/sglang/srt/models/deepseek_janus_pro.py:535]: change it into `global_server_args_dict`, the same as other vlm models

File: test/srt/models/test_vlm_models.py [hunk 11]
```diff
@@ -19,17 +28,22 @@
 
 # VLM models for testing
 MODELS = [
+    # SimpleNamespace(
+    #     model="google/gemma-3-4b-it", chat_template="gemma-it", mmmu_accuracy=0.384
+    # ),
+    # SimpleNamespace(
+    #     model="Qwen/Qwen2.5-VL-3B-Instruct",
```
Comment (yizhang2077) [inline test/srt/models/test_vlm_models.py]: remove comment here

File: python/sglang/srt/models/deepseek_janus_pro.py (hunk 10 above)
Comment (Othame) [reply python/sglang/srt/models/deepseek_janus_pro.py:535]: Actually, the real `qkv_backend` is determined within the vision attention implementation here https://github.com/sgl-project/sglang/blob/caea33ad1f1ee1d1b07b48e9359d567caecd4a9e/python/sglang/srt/layers/attention/vision.py#L350 
And it seems to be independent of the parameter `qkv_backend` passed in.

File: python/sglang/srt/layers/attention/vision.py [hunk 12]
```diff
@@ -366,3 +249,247 @@ def forward(
...
+            num_heads, world_size
+        )
+
+        self.q_size = self.num_attention_heads_per_partition * self.head_size
+        self.kv_size = self.num_attention_kv_heads_per_partition * self.head_size
+
+        if global_server_args_dict["mm_attention_backend"] is None:
+            if qkv_backend is None:
+                qkv_backend = "sdpa"
+            info_once(f"Multimodal attention backend not set. Use {qkv_backend} as default.")
```
Comment (mickqian) [inline python/sglang/srt/layers/attention/vision.py]: Can be changed to `Using {qkv_backend}.`, since not all scenarios in this case are default options

**Review by mickqian — APPROVED** [review]

**Comment by zhaochenyang20** [comment]
@Othame fix the lint and remining problems？

File: python/sglang/srt/layers/attention/vision.py (hunk 12 above)
Comment (Othame) [reply python/sglang/srt/layers/attention/vision.py]: OK！Change has been made.
